site/tests/app/controllers/admin/LibraryManageControllerTester.php

```diff
@@ -0,0 +1,147 @@
+            'status' => 'fail',
+            'message' => 'A file must be provided.'
+        ], $response->json);
+        $this->assertCount(0, $this->gateway->getAllLibraries($this->location));
+    }
+
+    /** @test */
+    public function testAjaxUploadLibraryFromZipSuccess() {
+        $_FILES['zip'] = [
+            'name' => 'lib.zip',
+            'tmp_name' => 'We all make mistakes in the heat of passion, jimbo.'
+        ];
+
+        $response = $this->controller->ajaxUploadLibraryFromZip()->json_response;
+
+        $this->assertInstanceOf(JsonResponse::class, $response);
+        $this->assertEquals([
+            'status' => 'success',
+            'data' => 'Successfully installed new library: lib'
+        ], $response->json);
+        $this->assertCount(1, $this->gateway->getAllLibraries($this->location));
+    }
+
+    /** @test */
+    public function testAjaxUploadLibraryFromGitSuccess() {
+        $_POST['git_url'] = 'https://github.com/Submitty/Submitty.git';
+        $response = $this->controller->ajaxUploadLibraryFromGit()->json_response;
+
+        $this->assertInstanceOf(JsonResponse::class, $response);
+        $this->assertEquals([
+            'status' => 'success',
+            'data' => 'Successfully cloned https://github.com/Submitty/Submitty.git.'
+        ], $response->json);
+        $this->assertCount(1, $this->gateway->getAllLibraries($this->location));
+    }
+
+    /** @test */
+    public function testAjaxUploadLibraryFromGitFail() {
+        $_POST['git_url'] = 'my first git url. BABY SHARK DO DO DO DO DO';
+
+        $response = $this->controller->ajaxUploadLibraryFromGit()->json_response;
+
+        $this->assertInstanceOf(JsonResponse::class, $response);
+        $this->assertEquals([
+            'status' => 'fail',
+            'message' => 'The git url is not of the right format.'
+        ], $response->json);
+        $this->assertCount(0, $this->gateway->getAllLibraries($this->location));
+    }
+
+    /** @test */
+    public function testItThrowsNotEnabledException() {
+        $this->expectException(NotEnabledException::class);
+        $this->createConfigWithLibrary(false);
+    }
+
+    /** @test */
+    public function testItDeletesAFileAfterHandling() {
+        $location = FileUtils::joinPaths(sys_get_temp_dir(), Utils::generateRandomString());
+        $this->createConfigWithLibrary(true, $location);
+        $this->assertTrue(FileUtils::createDir($this->location));
+        $file = $this->createFile('tmp.txt');
+
+        $_FILES['zip'] = [
+            'name' => 'lib.zip',
+            'tmp_name' => $file
+        ];
+
+        $this->controller->ajaxUploadLibraryFromZip()->json_response;
+
+        $this->assertFileNotExists($file);
+        $this->assertTrue(FileUtils::recursiveRmdir($this->location));
+    }
+
+    protected function createFile(string $name): string {
+        $name = FileUtils::joinPaths($this->location, $name);
+        touch($name);
```

Comment:
It seems that this [line would only work if invoked from a command line](https://www.php.net/manual/en/function.touch.php). 

I suggest adding \_\_DIR\_\_ as the first argument.

Reply:
Three things
 1. This is a test, when are you executing tests from the website?
 2. I don't want `__DIR__` here as the file I'm creating for testing needs to be in the special temp directory I created specifically for the test that gets deleted when the tests are done.
 3. I assume you are referring to relative file paths in which case if you look at the line right above (132) I am appending it with the full file path. So it can be run from any directory just fine without __DIR__. Perhaps I should use a different variable name to make that more clear.